Support deploying Ceph Jewel release (bsc#970510)

tserong opened the pull request:
The Ceph Jewel release runs the various ceph daemons as an unprivileged user and group (ceph:ceph), whereas previous releases ran everything as root. There's a bunch of stuff that needs changing in the barclamp to support this. I believe I've got most of it in this PR, but still need to test further.

Also, the changes here effectively _only_ work with Jewel (they rely on the ceph package having created the appropriate user accounts automatically), so if you use this code with an older Ceph release, it will almost certainly fail to deploy. I may be better off trying to make these changes conditionally depending on what version of Ceph is available, but I don't actually know how to check that from within the Chef recipes.

tserong committed:
Set directory owner to 'ceph' user

The Ceph Jewel release runs the various ceph daemons as the unprivileged
'ceph' user, so we need to set the directory permissions appropriately
(actually, it may be enough to drop the directory creation altogether
and leave it up to the ceph package...)

Signed-off-by: Tim Serong <tserong@suse.com>

tserong committed:
Enable ceph-{mon,osd} systemd targets

The Ceph Jewel release includes additional systemd targets that need to
be enabled in addition to the main ceph.target.

Signed-off-by: Tim Serong <tserong@suse.com>

tserong committed:
Set user and group when invoking ceph-mon --mkfs

Signed-off-by: Tim Serong <tserong@suse.com>

tserong commented:
radosgw deployment is presently broken

tserong commented:
The gating failure is because the ceph user doesn't exist (this is never going to succeed anyway until we've got the CI system using SES 3 with Jewel)

tserong committed:
Fix instance name and file perms for radosgw Jewel deployments (bsc#975882)

Signed-off-by: Tim Serong <tserong@suse.com>

AbelNavarro commented on chef/cookbooks/ceph/recipes/conf.rb:
```diff
@@ -17,16 +19,16 @@
-  mode "0755"
+  owner "ceph"
+  group "ceph"
+  mode "3770"
```
is that **sticky bit** and **setguid** necessary?

tserong replied:
I don't know why that's necessary offhand, but it matches what's in the ceph spec file:

`%attr(3770,ceph,ceph) %dir %{_localstatedir}/log/ceph/`

tserong committed:
Use "rgw." instead of "radosgw." for consistency with ceph-deploy

This also removes some unnecessary settings from ceph.conf which have
reasonable defaults (admin socket, log file, etc.)

Signed-off-by: Tim Serong <tserong@suse.com>

tserong committed:
Don't create unnecessary radosgw directories

Signed-off-by: Tim Serong <tserong@suse.com>

AbelNavarro commented:
triggered rebuild.

tserong commented:
From my perspective this is good to go, once https://github.com/SUSE-Cloud/automation/pull/991 is merged and the SES3 repos are available. Unless these changes need to be somehow made conditional on the version of Ceph that's available to install...? To be clear, if this PR is accepted as-is, Ceph >= Jewel becomes a requirement, i.e. it won't correctly deploy older Ceph versions.

vuntz commented on chef/cookbooks/ceph/recipes/conf.rb:
```diff
@@ -9,6 +9,8 @@
 end
 
+# These directories should all be created by the package, maybe we can
+# just remove this stuff...?  (same with the directories created in osd.rb)
```
It's unclear if they're needed for other distros, so one approach would be to only create them for non-SUSE platform families.

vuntz commented on chef/cookbooks/ceph/recipes/radosgw.rb:
```diff
@@ -68,6 +49,16 @@
+# In the systemd case, need extra targets enabled
+service "ceph-radosgw.target" do
+  action :enable
+  only_if { File.exist?("/usr/lib/systemd/system/ceph-radosgw.target") }
```
Why the `only_if`?

tserong replied:
mon.rb and osd.rb use `if service_type == "systemd"`, and `service_type` comes from `node["ceph"]["mon"]["init_style"]` or `node["ceph"]["osd"]["init_style"]`. There's no such equivalent attribute for rgw, so I went with `only_if` on the unit file instead.

vuntz replied:
Hrm, okay. I guess another way is to just put that inside a platform family check for "suse". But no big deal.

vuntz commented on chef/cookbooks/ceph/recipes/mon.rb:
```diff
@@ -91,7 +91,7 @@
 
   execute "ceph-mon mkfs" do
-    command "ceph-mon --mkfs -i #{node['hostname']} --keyring '#{keyring}'"
+    command "chown ceph:ceph #{keyring} ; ceph-mon --mkfs -i #{node['hostname']} --keyring '#{keyring}' --setuser ceph --setgroup ceph"
```
Is the `chown` required only when migrating from a previous release, or all the time? I find this slightly suboptimal, but I don't have a better proposal right now.

tserong replied:
The keyring needs to be owned by the ceph user so the daemons can read it, but (IIRC) because the earlier `ceph-authtool` invocations are being run by chef as root, the keyring ends up with root ownership by default.

tserong committed:
Don't create directories already owned by the Ceph pacakge

For now, this change is inside node["platform_family"] == "suse" guards
in case the Ceph package on other distros doesn't already create these
directories.

Signed-off-by: Tim Serong <tserong@suse.com>

vuntz commented:
+1, but needs to be merged with https://github.com/SUSE-Cloud/automation/pull/991 and https://github.com/crowbar/crowbar-core/pull/483
